app/src/main/java/org/gophillygo/app/di/AppModule.java

```diff
@@ -83,4 +90,12 @@ EventDao provideEventDao(GpgDatabase db) {
     AttractionFlagDao provideAttractionFlagDao(GpgDatabase db) {
         return db.attractionFlagDao();
     }
+
+    @Singleton
```

Comment:
So what's going on here is that we use Dagger for dependency injection (a dependency injection library called Dagger strikes me as...kind of grim). We need to create just a single instance of both `ExecutorService` and `Handler` at the top level of the app so that they're available to any parts of the app that need them, but the place where we need them, `DestinationRepository`, is nested many, many levels down from the top. If we did this by hand, we'd have to pass an instance of `ExecutorService` down through every nested constructor call in the application to get it down to the level where it's actually used. What Dagger does is the `@Singleton` decorator tells it to know about this resource and to make sure that only one of it ever exists, and the `@Provides` decorator tells it that if some other part of the app asks for one of these things, to get it from here. Then, in places where we need one or more parameters that Dagger knows about, we can add the `@Inject` decorator and Dagger will inspect the parameters to the constructor, compare them to the different resources that it knows about, and, assuming it finds a match, it'll auto-generate all the code to pass that resource all the way down to where it gets used so that we don't write it by hand. You can see this at work in [DestinationRepository.java](https://github.com/azavea/cac-tripplanner-android/pull/218/files#diff-f54b6a041c754c2c7047b1b20fa3d1b2a4e4f0be97acac6cfdd8f4f724406a59R60). I don't 100% understand what heuristics it uses to figure out which resource to attach to which constructor parameters (does it use the type signature alone or does it also parse the name somehow?) but it seems to be working, so that's good enough for now.

Reply:
This is really helpful context, thank you!